### PR TITLE
Check `assign_op_pattern` for conflicting borrows

### LINKED ISSUE #9180: assign_op_pattern suggested change is a compiler error for non-primitive vector elements
### Summary

In the repro case below, the borrow checker disallows a direct use of the assign operator when attempting to add two elements of a vector. I'm not sure what the necessary requirements are on the vector element types, but a wrapped `i32` suffices. If I split it out into the long form pattern, then clippy says to change it back even though that would be a compile error.

Specifically, in the repro below, clippy wants to change `x[0] = x[0] + x[1];` to `x[0] += x[1];`, but that results in the following compiler error:

```
  --> src/main.rs:22:13
   |
22 |     x[0] += x[1];
   |     --------^---
   |     |       |
   |     |       immutable borrow occurs here
   |     mutable borrow occurs here
   |     mutable borrow later used here
```

### Lint Name

assign_op_pattern

### Reproducer

I tried this code:

```rust
use core::ops::{Add, AddAssign};

#[derive(Debug, Copy, Clone)]
struct Foo(i32);

impl AddAssign for Foo {
    fn add_assign(&mut self, other: Self) {
        self.0 += other.0;
    }
}

impl Add for Foo {
    type Output = Self;

    fn add(self, other: Self) -> Self {
        Self(self.0+other.0)
    }
}

fn main() {
    let mut x = vec![Foo(1),Foo(2),Foo(3)];
    x[0] = x[0] + x[1];
    println!("{:?}", x[0]);
}
```

I saw this happen:

```
  --> src/main.rs:22:5
   |
22 |     x[0] = x[0] + x[1];
   |     ^^^^^^^^^^^^^^^^^^ help: replace it with: `x[0] += x[1]`
   |
   = note: `#[warn(clippy::assign_op_pattern)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#assign_op_pattern
```

I expected to see this happen:

No lint warning.

### Version

```text
I ran the minimal case above in the rust playground, with "Stable version: 1.62.0". I'm not sure how to get it to say the full version info.
```

### Additional Labels

_No response_

### PR DESCRIPTION
**Pull request by Jarcho**
fixes #9180

changelog: [`assign_op_pattern`](https://rust-lang.github.io/rust-clippy/master/#assign_op_pattern): Don't lint when the suggestion would cause borrowck errors.

### CODE REVIEW

**Review by Manishearth — APPROVED** [review]

File: clippy_lints/src/operators/assign_op_pattern.rs [hunk 1]
```diff
@@ -99,3 +113,69 @@ impl<'a, 'tcx> Visitor<'tcx> for ExprVisitor<'a, 'tcx> {
         walk_expr(self, expr);
     }
 }
+
+fn imm_borrows_in_expr(cx: &LateContext<'_>, e: &hir::Expr<'_>) -> hir::HirIdSet {
```
Comment (Manishearth) [inline clippy_lints/src/operators/assign_op_pattern.rs:117]: thought: should these live in utils?

also worth adding a doc comment

**Review by Manishearth — APPROVED** [review]
r=me either way

File: clippy_lints/src/operators/assign_op_pattern.rs (hunk 1 above)
Comment (Jarcho) [reply clippy_lints/src/operators/assign_op_pattern.rs:117]: I would imagine anything else would need to gather a list of all borrows, not just mutable or immutable ones.

Comment (Manishearth) [reply clippy_lints/src/operators/assign_op_pattern.rs:117]: ah fair

**Pull request merged**